### ISSUE TITLE
kv: panic in HandleRaftUncoalescedRequest in mixed cluster

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#128747]
While working on adding 24.3 versions (#128616), I noticed the `version-upgrade` roachtest failing, with one node crashing here:
```
runtime error: index out of range [24] with length 24
*   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleRaftUncoalescedRequest
```
Code:
```go
	s.metrics.RaftRcvdMessages[req.Message.Type].Inc(1)
```

Two new message types were added after 24.2 branched in #128426

~Looks like this problem is in 24.1 as well, which will be problematic in 24.1 -> 24.3 upgrades.~ Edit: actually, if this was true, we would have seen failures on upgrades test on master and on 24.2. To be investigated some more

Jira issue: CRDB-41178

**Comment by RaduBerinde** [comment]
The solution here might be to put the use of these new message type behind version gates (in which case we don't need to fix anything in 24.2).

**Comment by RaduBerinde** [comment]
Ok, I got to the bottom of this. #128647 added a pre-24.2 version gate to master.
